### DISCUSSION TITLE
Reasoning for Library usage vs Roslyn directly?

### DISCUSSION THREAD

**Discussion by matt-psaltis** [microsoft/RulesEngine#268] [General]
This is purely a curiosity question. What drives the team to choose FastExpressionCompiler and System.Linq.Dynamic.Core vs using Roslyn directly?

Is there some fundamental concerns which make Roslyn a "bad" choice for this scenario? Is it purely historical/timing related?

Cheers!

**Comment by abbasc52** [comment]
We evaluated Rosyln as first choice but came across these limitations:
- Rosyln requires the type of globals (or Input parameters) as a strong typed class which should be available in an assembly. Since we wanted to support generic inputs, Rosyln can not support it.
- Rosyln is more that just an expression compiler, It supports full compilation of c# code and we found that the performance was very slow when using it at runtime.

Regarding `System.Linq.Dynamic.Core` we used it because it was the most complete C# expression library we could find. Along with the ability to inject custom classes and types.

`FastExpressionCompiler` was added later as a performance optimization. As per our benchmarks, it improved the performance by 35%
